packages/@lwc/engine-server/src/renderer.ts

```diff
@@ -330,23 +338,114 @@ function isConnected(node: HostNode) {
 // synthetic shadow.
 const insertStylesheet = noop as (content: string, target: any) => void;
 
-// Noop on SSR.
-const addEventListener = noop as (
-    target: HostNode,
+function addEventListener(
+    node: N,
     type: string,
     callback: EventListener,
-    options?: AddEventListenerOptions | boolean
-) => void;
+    useCaptureOrOptions?: AddEventListenerOptions | boolean
+): void {
+    if (node.type !== 'element') {
+        return;
+    }
+
+    let savedCallback = callback;
+
+    if (useCaptureOrOptions) {
+        if (useCaptureOrOptions === true || useCaptureOrOptions.capture) {
+            // Capturing listeners aren't meaningful in SSR, since the capture phase
+            // is not supported. However, we don't need to break apps that register
+            // these listeners.
+            return;
+        }
+        if (useCaptureOrOptions.once) {
+            savedCallback = (...args) => {
+                try {
+                    callback(...args);
+                } finally {
+                    removeEventListener(node, type, savedCallback);
+                }
+            };
+        }
+
+        // In the SSR context, `passive` event listeners are not meaningful and are ignored.
+        // Additionally, we do not currently support AbortSignals, so the `signal` option
+        // is also ignored.
+    }
 
-// Noop on SSR.
-const removeEventListener = noop as (
-    target: HostNode,
+    if (!(type in node.eventListeners)) {
+        node.eventListeners[type] = new Set();
```

Comment:
I would suggest we use a `Map` here for `eventListeners` instead of an object. Otherwise weird things can happen if a developer does `node.addEventListener('constructor', ...)` or `node.addEventListener('__proto__', ...)`.

Probably it's not vulnerable to any kind of security issue (e.g. prototype pollution), but why take the risk? 🙂